smac/intensification/intensification.py

```diff
@@ -137,94 +190,362 @@ def __init__(self, tae_runner: ExecuteTARun,
         if hasattr(challenger, 'origin'):
             self.logger.debug("Configuration origin: %s", challenger.origin)
 
-        try:
-            # since it runs only 1 "ExecuteRun" per iteration,
-            # we run incumbent once and then challenger in the next
-            if self.stage == IntensifierStage.RUN_FIRST_CONFIG or \
-                    self.stage == IntensifierStage.RUN_INCUMBENT:
-                # Lines 3-7
-                self._add_inc_run(incumbent=incumbent, run_history=run_history, log_traj=log_traj)
-            elif self.stage == IntensifierStage.RUN_CHALLENGER or \
-                    self.stage == IntensifierStage.RUN_BASIS:
-                # Lines 8-17
-                incumbent = self._race_challenger(challenger=challenger,
-                                                  incumbent=incumbent,
-                                                  run_history=run_history,
-                                                  log_traj=log_traj)
-
-        except BudgetExhaustedException:
-            # We return incumbent, SMBO stops due to its own budget checks
-            inc_perf = run_history.get_cost(incumbent)
-            self.logger.debug("Budget exhausted; Return incumbent")
-            return incumbent, inc_perf
-
-        # time used to evaluate challenger
-        self.elapsed_time += time.time() - start_time
+        if self.stage in [IntensifierStage.RUN_CHALLENGER, IntensifierStage.RUN_BASIS]:
+
+            if not self.to_run:
+                self.to_run, self.inc_sum_cost = self._get_instances_to_run(
+                    incumbent=incumbent,
+                    challenger=challenger,
+                    run_history=run_history,
+                    N=self.N
+                )
 
+            is_there_time_due_to_adaptive_cap = self._is_there_time_due_to_adaptive_cap(
+                challenger=challenger,
+                run_history=run_history,
+            )
+
+            # If there is no more configs to run in this iteration, or no more
+            # time to do so, change the current stage base on how the current
+            # challenger performs as compared to the incumbent. This is done
+            # via _process_racer_results
+            if len(self.to_run) == 0 or not is_there_time_due_to_adaptive_cap:
+
+                # If no more time, stage transition is a must
+                if not is_there_time_due_to_adaptive_cap:
+                    self.stage = IntensifierStage.RUN_INCUMBENT
+                    self.logger.debug(
+                        "Stop challenger itensification due "
+                        "to adaptive capping."
+                    )
+
+                # Nevertheless, if there are no more instances to run,
+                # we might need to comply with line 17 and keep running the
+                # same challenger
+                self.logger.debug("No further runs for challenger possible")
+                self._process_racer_results(
```

Comment:
I have the same questions here as with the function process incumbent runs. Is this really necessary and shouldn't such handling be done in the process_results?

Reply:
I think the justification for this to be like that (as in the previous code) is that if there is no new challenger, you don't call process results, so this piece of code makes sure that on the next iteration we move to run the incumbent or we keep running the challenger, based on the run_history.

Reply:
Okay, got it now. Could you please add more info that line 17 actually doubles the number of instances to be compared here?

Reply:
thanks, I updated the full comment to be:
                   # Nevertheless, if there are no more instances to run,
                   # we might need to comply with line 17 and keep running the
                   # same challenger. In this case, if there is not enough information
                   # to decide if the challenger is better/worst than the incumbent,
                   # line 17 doubles the number of instances to run.